src/libopensc/card-myeid.c

```diff
@@ -123,27 +114,14 @@ static int myeid_get_card_caps(struct sc_card *card, myeid_card_caps_t* card_cap
 
 static int myeid_match_card(struct sc_card *card)
 {
-	int i, match = -1;
-
-	for (i = 0; myeid_atrs[i] != NULL; i++)
-	{
-		u8 defatr[SC_MAX_ATR_SIZE];
-		size_t len = sizeof(defatr);
-		const char *atrp = myeid_atrs[i];
-
-		if (sc_hex_to_bin(atrp, defatr, &len))
-			continue;
-		if (len != card->atr.len)
-			continue;
-		if (memcmp(card->atr.value, defatr, len) != 0)
-			continue;
-		match = i;
-		break;
+	/* Normally the historical bytes are exactly "MyEID", but there might
+	 * be some historic units which have a small prefix byte sequence. */
+	if (card->reader->atr_info.hist_bytes_len >= 5 &&
+	    !memcmp(&card->reader->atr_info.hist_bytes[card->reader->atr_info.hist_bytes_len - 5], "MyEID", 5)) {
+		card->type = SC_CARD_TYPE_MYEID_GENERIC;
+		return 1;
```

Comment:
You may want to register your card for Windows' minidriver via `win32/customactions.cpp`.

Reply:
Thanks. If done it's worth separate commit. I let @hhonkanen to figure out this as separate step if needed.

Reply:
In a similar case, I've decided to simply list all possible ATRs with an ATR mask:
https://github.com/OpenSC/OpenSC/blob/46930fc0c46a41c4a2e0121d7012cab6030f0dfd/src/libopensc/card-sc-hsm.c#L62-L131
and the corresonding code in the Windows installer:
https://github.com/OpenSC/OpenSC/blob/46930fc0c46a41c4a2e0121d7012cab6030f0dfd/win32/customactions.cpp#L77-L100

Reply:
I am not sure about registering for Windows minidriver. Aventra maintains a MyEID specific minidriver, which is tested using Microsoft Hardware Lab Kit tools upon each release. I don't have experience on using MyEID with OpenSC's minidriver and we ususally recommend our customers to use our MyEID Minidriver for CSP/CNG support on Windows, and OpenSC for PKCS#11. Because I don't know what OpenSC's minidriver is like compared to our own, I think it is better to consider this separately, and not within this PR. 

If somebody has installed our minidriver and then installs OpenSC for PKCS#11, could it accidentally override the previous minidriver installation? Is the minidriver option enabled by default in OpenSC's Windows installer?

Reply:
> In a similar case, I've decided to simply list all possible ATRs with an ATR mask:
> 
> [OpenSC/src/libopensc/card-sc-hsm.c](https://github.com/OpenSC/OpenSC/blob/46930fc0c46a41c4a2e0121d7012cab6030f0dfd/src/libopensc/card-sc-hsm.c#L62-L131)

Other card drivers e.g. `card-piv.c` prefer historical bytes matching. As it's wider match and simpler code I chose this approach.

Seems there's some drivers that first try full ATR match and fall back to historical bytes matching. And some that do just ATR matching. So there's a variety of ways to do it in the tree.

If preferred I'm happy to convert the code to use instead _sc_match_atr(). Perhaps it would be good to document the preference somewhere? Or maybe I missed it?

Reply:
In regard to PIV driver: Windows provides a builtin PIV driver, based on looking for the PIV AID. 

All PIV cards must respond to SELECT AID for the PIV AID.  There are many PIV  NIST approved cards and PIV look-a-like applets on various cards/tokens. Many of these don't implement required features correctly or add features.  Thus much of the OpenSC PIV match and init code deals with determining if this card/token could have multiple applets, is an approved or look-a-like card. Some of this can be determined from known ATRs or if the Historical bytes contain the PIV AID. 

Look-a-like applet/card vendors write their applets to work with Windows. So most people will use the windows driver. Thus not much testing has been done with the OpenSC minidriver with PIV.